pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+            };
+
+            let build_and_earn_eras_per_cycle =
+                match T::CycleConfiguration::build_and_earn_eras_per_cycle() {
+                    0 => Balance::MAX,
+                    build_and_earn_eras_per_cycle => Balance::from(build_and_earn_eras_per_cycle),
+                };
+
+            let periods_per_cycle = match T::CycleConfiguration::periods_per_cycle() {
+                0 => Balance::MAX,
+                periods_per_cycle => Balance::from(periods_per_cycle),
+            };
+
+            // 3.1. Collator & Treausry rewards per block
+            let collator_reward_per_block = collators_emission / blocks_per_cycle;
+            let treasury_reward_per_block = treasury_emission / blocks_per_cycle;
+
+            // 3.2. dApp reward pool per era
+            let dapp_reward_pool_per_era = dapps_emission / build_and_earn_eras_per_cycle;
+
+            // 3.3. Staking reward pools per era
+            let base_staker_reward_pool_per_era =
+                base_stakers_emission / build_and_earn_eras_per_cycle;
+            let adjustable_staker_reward_pool_per_era =
+                adjustable_stakers_emission / build_and_earn_eras_per_cycle;
+
+            // 3.4. Bonus reward pool per period
+            let bonus_reward_pool_per_period = bonus_emission / periods_per_cycle;
+
+            // 4. Block at which the inflation must be recalculated.
+            let recalculation_block = now.saturating_add(T::CycleConfiguration::blocks_per_cycle());
+
+            // 5. Return calculated values
+            (
+                max_emission,
+                InflationConfiguration {
+                    recalculation_block,
+                    collator_reward_per_block,
+                    treasury_reward_per_block,
+                    dapp_reward_pool_per_era,
+                    base_staker_reward_pool_per_era,
+                    adjustable_staker_reward_pool_per_era,
+                    bonus_reward_pool_per_period,
+                    ideal_staking_rate: params.ideal_staking_rate,
+                },
+            )
+        }
+    }
+
+    impl<T: Config> StakingRewardHandler<T::AccountId> for Pallet<T> {
+        fn staker_and_dapp_reward_pools(total_value_staked: Balance) -> (Balance, Balance) {
+            let config = ActiveInflationConfig::<T>::get();
+            let total_issuance = T::Currency::total_issuance();
+
+            // First calculate the adjustable part of the staker reward pool, according to formula:
+            // adjustable_part = max_adjustable_part * min(1, total_staked_percent / ideal_staked_percent)
+            // (These operations are overflow & zero-division safe)
+            let staked_ratio = Perquintill::from_rational(total_value_staked, total_issuance);
+            let adjustment_factor = staked_ratio / config.ideal_staking_rate;
+
+            let adjustable_part = adjustment_factor * config.adjustable_staker_reward_pool_per_era;
```

Comment:
Not sure if I missed anything, but where is the `min(1, ...)` operation mentioned in comment?

Reply:
`let staked_ratio = Perquintill::from_rational(total_value_staked, total_issuance);`

is saturating, hence the `min(1, ...)` is implicit.